include/ofi_mr.h

```diff
@@ -0,0 +1,116 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+
+#include "fi.h"
+#include "fi_util.h"
+
+/*
+ * MR monitor
+ */
+
+#define UTIL_MR_MONITOR_GET_PROV(mr_monitor)	((mr_monitor)->attr.prov)
+
+struct ofi_mr_monitor_attr {
+	struct fi_provider	*prov;
+
+	int (*init)(void);
+	int (*open)(void **notifier);
+	int (*close)(void *notifier);
+	int (*subscribe)(void *notifier, void *addr,
+			 uint64_t len, uint64_t cookie);
+	int (*unsubscribe)(void *notifier, uint64_t cookie);
+	int (*get_event)(void *notifier, void *buf, size_t len);
+};
+struct ofi_mr_monitor {
+	fastlock_t			lock;
+	struct ofi_mr_monitor_attr	attr;
+	void				*notifier;
+	struct dlist_entry		rq_head;
+	int				ref_cnt;
+};
+
+struct ofi_mr_monitor_rq {
+	fastlock_t		lock;
+	struct dlist_entry	list;
+	struct dlist_entry	entry;
+};
+
+struct ofi_mr_monitor_context {
+	struct ofi_mr_monitor_rq	*rq;
+	uint64_t			cookie;
+	struct dlist_entry		entry;
+};
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor_attr *attr);
+int ofi_mr_monitor_open(struct ofi_mr_monitor **mr_monitor);
+int ofi_mr_monitor_close(struct ofi_mr_monitor *mr_monitor);
+int ofi_mr_monitor_get_event(struct ofi_mr_monitor *mr_monitor,
+			     struct ofi_mr_monitor_rq *rq,
+			     struct ofi_mr_monitor_context **context);
+static inline
+int ofi_mr_monitor_subscribe(struct ofi_mr_monitor *mr_monitor,
+			     struct ofi_mr_monitor_rq *rq,
+			     void *addr, uint64_t len,
+			     uint64_t cookie,
+			     struct ofi_mr_monitor_context *context)
+{
+	int ret;
+
+	assert(context && rq && mr_monitor);
+	context->rq = rq;
+	context->cookie = cookie;
+
+	ret = mr_monitor->attr.subscribe(mr_monitor->notifier, addr,
+					 len, (uint64_t)context);
+	if (!ret)
+		FI_DBG(UTIL_MR_MONITOR_GET_PROV(mr_monitor), FI_LOG_MR,
+		       "monitoring addr=%p len=%"PRIu64" cookie=%"PRIu64
+		       "context=%p rq=%p notifier=%p\n",
+		       addr, len, context->cookie,
+		       context, rq, mr_monitor->notifier);
+	return ret;
+}
+static inline
+int ofi_mr_monitor_unsubscribe(struct ofi_mr_monitor *mr_monitor,
+			       uint64_t cookie,
+			       struct ofi_mr_monitor_context *context)
+{
+	assert(mr_monitor && (cookie == context->cookie));
+	return mr_monitor->attr.unsubscribe(mr_monitor->notifier,
+					    (uint64_t)context);
+}
```

Comment:
There's a chance that the subscription is queued onto an event list.  After unsubscribe, we need to do something like this:
```
fastlock_acquire(&eq->lock);
dlist_remove(&subscript->entry);
fastlock_release(&eq->lock);
```

Reply:
Done